File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/ast/early_return.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid early return rule."""

import sys

from collections.abc import Iterable

import libcst as cst

from rules.harness_check_rule import (
    Finding,
    FindingFix,
    FixSafety,
    HarnessCheckRule,
    has_nested_function,
    parse_python,
)
from core.rule_context import RuleContext, relative, stack_sources_configured

sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")


class EarlyReturnRule(HarnessCheckRule):
    """Validate earlyReturn check."""

    category = "earlyReturn"

    def applies(self, ctx: RuleContext) -> bool:
        """Check if this rule applies to the context."""
        if not ctx.is_enabled(self.category):
            return False
        return stack_sources_configured(ctx.manifest.raw, self.category)

    def validate(self, ctx: RuleContext) -> Iterable[Finding]:
        """Validate earlyReturn check."""
        severity = ctx.severity_of(self.category)
        category = self.category
        sources = ctx.stack_sources(self.category)

        class EarlyReturnFinder(cst.CSTVisitor):
            METADATA_DEPENDENCIES = (cst.metadata.PositionProvider,)

            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path

            def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
                if has_nested_function(node):
                    return False
                if not isinstance(node.body, cst.IndentedBlock):
                    return True
                if not node.body.body:
                    return True
                for stmt in node.body.body[:-1]:
                    if isinstance(stmt, cst.SimpleStatementLine):
                        for inner_stmt in stmt.body:
                            if isinstance(inner_stmt, cst.Return):
                                pos = self.get_metadata(
                                    cst.metadata.PositionProvider, stmt
                                )
                                self.findings.append(
                                    Finding(
                                        severity,
                                        category,
                                        f"{self.rel_path}:{pos.start.line}: function `{node.name.value}` has an early/mid return; restructure with single exit",
                                        file=self.rel_path,
                                        start_line=pos.start.line,
                                        start_column=pos.start.column + 1,
                                        end_line=pos.end.line,
                                        end_column=pos.end.column + 1,
                                        fix=FindingFix(
                                            description="rewrite with early return; review nested logic before applying",
                                            safety=FixSafety.UNSAFE,
                                            edits=(),
                                        ),
                                    )
                                )
                return True

        def collect_findings():
            for path in sources:
                tree, error = parse_python(path)
                if error is not None:
                    yield Finding(
                        severity,
                        category,
                        f"{relative(path, ctx.root)}: syntax error: {error}",
                    )
                    continue
                wrapper = cst.MetadataWrapper(tree)
                visitor = EarlyReturnFinder(relative(path, ctx.root))
                wrapper.visit(visitor)
                yield from visitor.findings
        return list(collect_findings())


RULE: HarnessCheckRule = EarlyReturnRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/ast/early-return.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type { FunctionLike, Node, SourceFile, Statement } from "typescript@6.0.3";
import {
    createSourceFile,
    isBlock,
    isFunctionDeclaration,
    isIdentifier,
    isMethodDeclaration,
    isReturnStatement,
    SyntaxKind,
} from "typescript@6.0.3";
import { astChildrenOf } from "../../core/ast-traversal";
import type { Finding, HarnessCheckRule, RuleContext } from "../harness-check-rule";

/**
 * Forbid early return statements in functions.
 */
export const earlyReturnRule: HarnessCheckRule = {
    category: "earlyReturn",
    applies(_: RuleContext): boolean {
        return true;
    },

    validate(ctx: RuleContext): readonly Finding[] {
        return ctx
            .stackSources("earlyReturn")
            .map((file) => ({ file, text: ctx.read(file) }))
            .filter(({ text }) => text !== "")
            .flatMap(({ file, text }) => {
                const sourceFile: SourceFile = createSourceFile(file, text, SyntaxKind.LatestVersion, true);
                const getFuncName = (funcNode: FunctionLike): string => {
                    if (isFunctionDeclaration(funcNode) && funcNode.name) {
                        return funcNode.name.text;
                    }
                    if (isMethodDeclaration(funcNode) && funcNode.name && isIdentifier(funcNode.name)) {
                        return funcNode.name.text;
                    }
                    return "<anonymous>";
                };
                const visitNode = (node: Node): readonly Finding[] => {
                    switch (node.kind) {
                        case SyntaxKind.FunctionDeclaration:
                        case SyntaxKind.MethodDeclaration:
                        case SyntaxKind.FunctionExpression:
                        case SyntaxKind.ArrowFunction:
                        case SyntaxKind.Constructor: {
                            const funcNode = node as FunctionLike;
                            if (!funcNode.body || !isBlock(funcNode.body) || funcNode.body.statements.length === 0) {
                                return astChildrenOf(node).flatMap(visitNode);
                            }
                            const earlyReturns = funcNode.body.statements
                                .slice(0, -1)
                                .filter(isReturnStatement)
                                .map((stmt: Statement) => {
                                    const start = sourceFile.getLineAndCharacterOfPosition(stmt.getStart(sourceFile));
                                    const end = sourceFile.getLineAndCharacterOfPosition(stmt.getEnd());
                                    return {
                                        severity: ctx.severityOf("earlyReturn"),
                                        category: "earlyReturn",
                                        message: `function \`${getFuncName(funcNode)}\` has an early return; restructure with single exit`,
                                        file,
                                        startLine: start.line + 1,
                                        startColumn: start.character + 1,
                                        endLine: end.line + 1,
                                        endColumn: end.character + 1,
                                        fix: {
                                            description: "restructure function with single exit point",
                                            safety: "unsafe",
                                            edits: [],
                                        },
                                    };
                                });
                            return earlyReturns.concat(astChildrenOf(node).flatMap(visitNode));
                        }
                        default:
                            return astChildrenOf(node).flatMap(visitNode);
                    }
                };
                return visitNode(sourceFile);
            });
    },
};
